### DISCUSSION TITLE
What if we have a block scoped "try" like we have now for "using var"?

### DISCUSSION THREAD

**Discussion by VyacheslavGarshin** [dotnet/csharplang#6886] [Language Ideas] [answered]
Currently "try" operator makes another level of nesting and in case it wraps the whole method than the function body look a bit awkward.
Wouldn't it be beautiful to have?:
```
void DoStuff() 
{
    try;

    // doing stuff here
    var stuff1 = OtherStuff();
    var stuff2 = OtherStuff2();

    catch(Exception ex);
    // deal with exception

    finally;
    // finally body
}
```

**Comment by theunrepentantgeek** [comment, marked as answer]
The code would be ambiguous.

Consider this slightly larger example:

``` csharp
void DoStuff() 
{
    try;

    // doing stuff here
    var stuff1 = OtherStuff();

    try;

    // doing more stuff here
    var stuff2 = OtherStuff2();

    catch (InvalidOperationException ex);
    // deal with Invalid Operation exception

    catch (FileOperationException ex);
    // deal with File Operation exception

    finally;
    // finally body
}
```

What does this mean?

It might mean this:

``` csharp
void DoStuff() 
{
    try
    {
        // doing stuff here
        var stuff1 = OtherStuff();

        try
        {
            // doing more stuff here
            var stuff2 = OtherStuff2();
        }
        catch (InvalidOperationException ex)
        {
            // deal with Invalid Operation exception
        }
        catch (FileOperationException ex)
        {
            // deal with File Operation exception
        }
    }
    finally
    {
        // finally body
    }
}
```

or it might mean this:

``` csharp
void DoStuff() 
{
    try
    {
        // doing stuff here
        var stuff1 = OtherStuff();

        try
        {
            // doing more stuff here
            var stuff2 = OtherStuff2();
        }
        catch (InvalidOperationException ex)
        {
            // deal with Invalid Operation exception
        }
    }
    catch (FileOperationException ex)
    {
        // deal with File Operation exception
    }
    finally
    {
        // finally body
    }
}
```

Blocks and appropriate indentation aren't evil, they're clear - and clarity is vitally important to the goal of writing correct code.

Making code needlessly hug the lefthand margin isn't beautiful - it's cryptic and obfuscated.

**Reply by VyacheslavGarshin** [reply]
Yes. Thought about this. 
But if we limit block level "try" by only one occurrence then there will not be chance to misinterpret the statement. 
Just not sure if such limitation is acceptable and doesn't break language patterns. Can't recall such limitations in other statements.
Valid nested try could be:
```
void DoStuff() 
{
    try;

    // doing stuff here
    var stuff1 = OtherStuff();

    try
    {
        // doing more stuff here
        var stuff2 = OtherStuff2();
    }
    catch (InvalidOperationException ex)
    {
        // deal with Invalid Operation exception
    }

    catch (FileOperationException ex);
    // deal with File Operation exception

    finally;
    // finally body
}
```
PS/ Please, let's not discuss beauty of the code=) It's matter of personal taste. I like less nesting, others don't...

**Reply by DavidArno** [reply]
![hadouken indentation](https://external-preview.redd.it/6TTm4tS5pMCiq2KlgviubpGKt1QzvetOi2Z6Z3lB1Nc.jpg?auto=webp&v=enabled&s=3a7959b9421d5de0ef58784f1ed6fee8006bfb65)

No one should like writing deeply nested "hadouken code" 😆

**Reply by theunrepentantgeek** [reply]
> Please, let's not discuss beauty of the code=) It's matter of personal taste. I like less nesting, others don't...

I only brought it up because of your original comment:

> Wouldn't it be beautiful to have?

If you're going to argue that a language change should occur just because it makes code beautiful, you should expect people to respond to that.

I thought of something else though.

Indentation reflects the scoped structure of the code - the `{` braces `}` of the existing **try..catch..finally** syntax clearly delineate the scope across which the additional semantics of the **try** apply.

With _implicit_ **try;**, that scope _still exists_ - there's still a block of code subject to the additional semantics of running within a **try**. 

But now, that scope isn't reflected in the structure of the code, forcing any dev reading the code to carry that context in their head.

This _increases_ the cognitive load of a developer trying to read and understand. 
It's objectively and measurably just a little harder to comprehend.

**Reply by VyacheslavGarshin** [reply]
I raised the questions because we have already "using var". It does all that you've mentioned. Thought it would make sense to broaden this practice.

**Reply by theunrepentantgeek** [reply]
With **using var** a dev reading the code understands everything they need to know by reading that one line: the contained object will be disposed at end of the current scope (usually the entire method). This matches pre-existing usage - research prior to the adoption of **using var** had shown that the overwhelming majority of `using ( )` blocks finished at the end of the containing block. 

Because the interesing bits of **try..finally** and **try..catch** blocks are at the end, developers must read ahead to interpret the **finally** or the **catch** to understand the authors intent. Fortunately, these are easy to find because intervening code is (almost) always indented out of the way.

With the proposed syntax for **try;** the same forward scan would be required, but now their eyes can't saccade to the right spot instantly, they'd be forced to skim all the intervening code to try and find it. This would be much slower.

**Reply by HaloFour** [reply]
I think the other big thing with `using` that was more a quirk of that syntax is that if you needed to use a few different resources of different types you needed a separate block for each.  It was common to see ADO.NET code requiring 2-3 levels of indentation.  If `using` didn't have that limitation perhaps `using` variables would have been much less beneficial.
